File: neetcode150/medium/49-Group-Anagrams.py
#!/usr/bin/python3
class Solution:
  # Initial Solution
  def groupAnagrams(self, strs: list) -> list:
    g = []
    for w in strs:
      s = sorted(w)
      s = ''.join(s)
      g.append(s)
    
    h = {}
    for i, w in enumerate(g):
      if w in h:
        h[w].append(strs[i])
      else:
        h[w] = [strs[i]]
    
    sol = []
    for s in set(g):
      sol.append(h[s])

    return sol
  
  # Cleaner solution
  def groupAnagrams(self, strs: list) -> list:
    d = {}
    for w in strs:
      s = ''.join(sorted(w))
      if s not in d:
        d[s] = [w]
      else:
        d[s] += [w]
    # Two ways to return a dictionaries values
    # return [d[i] for i in d]
    return d.values()

File: neetcode150/medium/49-Group-Anagrams.cpp
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

// Solution 1
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> h;
        for(int i = 0; i < strs.size(); i++) {
            string val = strs[i];
            sort(val.begin(), val.end());
            h[val].push_back(strs[i]);
        }
        vector<vector<string>> sol;
        for(auto pair : h) {
            sol.push_back(pair.second);
        }
        return sol;
    }
};